src/recap.c

```diff
@@ -0,0 +1,490 @@
+      os_free(data);
+      return -1;
+    }
+    os_memcpy(&pctx->pcap_data[pctx->data_size], data, read_size);
+    pctx->data_size = current_size;
+  }
+  os_free(data);
+
+  return read_size;
+}
+
+int process_pcap_header_state(struct pcap_stream_context *pctx) {
+  ssize_t read_size = 0;
+  ssize_t pcap_header_size = (ssize_t)sizeof(struct pcap_file_header);
+
+  size_t len = (pcap_header_size > pctx->data_size)
+                   ? pcap_header_size - pctx->data_size
+                   : 0;
+
+  if ((read_size = read_pcap(pctx, len)) < 0) {
+    log_error("read_pcap fail");
+    return -1;
+  }
+
+  if (pctx->data_size >= pcap_header_size) {
+    log_trace("Received pcap header:");
+    os_memcpy(&pctx->pcap_header, pctx->pcap_data, pcap_header_size);
+    log_trace("\tpcap_file_header version_major = %d",
+              pctx->pcap_header.version_major);
+    log_trace("\tpcap_file_header version_minor = %d",
+              pctx->pcap_header.version_minor);
+    log_trace("\tpcap_file_header snaplen = %d", pctx->pcap_header.snaplen);
+    log_trace("\tpcap_file_header linktype = %d", pctx->pcap_header.linktype);
+    pctx->data_size = 0;
+    pctx->state = PCAP_STATE_READ_PKT_HEADER;
+  } else if (read_size == 0) {
+    log_trace("No data received");
+    pctx->state = PCAP_STATE_FIN;
+  }
+
+  return 0;
+}
+
+int process_pkt_header_state(struct pcap_stream_context *pctx) {
+  ssize_t read_size = 0;
+  ssize_t pkt_header_size = (ssize_t)sizeof(struct pcap_pkthdr32);
+
+  size_t len = (pkt_header_size > pctx->data_size)
+                   ? pkt_header_size - pctx->data_size
+                   : 0;
+
+  if ((read_size = read_pcap(pctx, len)) < 0) {
+    log_error("read_pcap fail");
+    return -1;
+  }
+
+  if (pctx->data_size >= pkt_header_size) {
+    log_trace("Received pkt header:");
+    os_memcpy(&pctx->pkt_header, pctx->pcap_data, pkt_header_size);
+    log_trace("\tpcap_pkthdr ts_sec = %llu", pctx->pkt_header.ts_sec);
+    log_trace("\tpcap_pkthdr ts_usec = %llu", pctx->pkt_header.ts_usec);
+    log_trace("\tpcap_pkthdr caplen = %llu", pctx->pkt_header.caplen);
+    log_trace("\tpcap_pkthdr len = %llu", pctx->pkt_header.len);
+
+    if (pctx->pkt_header.caplen > pctx->pkt_header.len) {
+      log_error("caplen > len");
+      return -1;
+    }
+
+    pctx->data_size = 0;
+    pctx->state = PCAP_STATE_READ_PACKET;
+  } else if (read_size == 0) {
+    log_trace("No data received");
+    pctx->state = PCAP_STATE_FIN;
+  }
+
+  return 0;
+}
+
+void get_packet_header(struct pcap_stream_context *pctx,
+                       struct pcap_pkthdr *header) {
+  header->ts.tv_sec = pctx->pkt_header.ts_sec;
+  header->ts.tv_usec = pctx->pkt_header.ts_usec;
+  header->caplen = pctx->pkt_header.caplen;
+  header->len = pctx->pkt_header.len;
+}
+
+void free_packets(UT_array *packets) {
+  struct tuple_packet *p = NULL;
+  while ((p = (struct tuple_packet *)utarray_next(packets, p)) != NULL) {
+    free_packet_tuple(p);
+  }
+}
```

Comment:
Instead of using this function, just set a `.dtor_f` (destructor function) in your `UT_Array` `UT_icd` struct, see https://troydhanson.github.io/uthash/utarray.html#_about_ut_icd:

```c
static const UT_icd tp_list_icd = {sizeof(struct tuple_packet), NULL, NULL,
                                   .dtor_f = free_packet_tuple};
```

Reply:
Very good find. We will have add this destructor everywhere in the edgesec code.